keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -275,20 +275,57 @@ with a metric.
 
 There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade -- that would cause other problems, too.
 
-TODO: security / cert stuff.
+We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve).
+
+To prevent Man in the Middle attack we will not expose information about apiserver IP/endpoint and the trust bundle control (used to authenticate server while proxying) in REST APIs.
 
 ## Design Details
 
-TODO: explanation of how the handler will determine a request is for a resource
-that should be proxied.
+### Aggregation Layer
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource
+  1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+  1. It is a non resource request
+  2. The StorageVersion informer cache hasn't synced yet or if `StorageVersionManager.Completed()` has returned false
+  3. The request has a header that indicates that this request has been proxied once already
+  4. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+  5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `No relevant API server found for the requested GVR <group_version_resource>`
```

Comment:
This case shouldn't happen, but I suggest the message should be "GVR is not served by anything in this cluster", i.e. it's a fact about the cluster, not an error condition about not finding an apiserver.